controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -295,6 +308,66 @@ public void getCurrentSegments(StreamInfo request, StreamObserver<SegmentRanges>
                 responseObserver);
     }
 
+    private Supplier<String> createDelegationTokenSupplier(StreamInfo request) {
+        return () -> {
+            StreamAuthParams authParams = new StreamAuthParams(request.getScope(), request.getStream(),
+                    this.isInternalWritesWithReadPermEnabled);
+
+            // StreamResource will take on values containing `_RG` (ex: "prn:://scope:myScope/stream:_RGmyApp")
+            // and `_MARK` for internal streams. We use the same strings in delegation tokens sent to Segment Store,
+            // while we use resource tags lie "reader-group" (ex: "prn:://scope:myScope/reader-group:myApp)
+            // and "watermark" for authorization in Controller.
+            String streamResource = authParams.streamResourceString();
+
+            // This is for backward compatibility.
+            if (request.getRequestedPermission().equals("") && authParams.isStreamUserDefined()) {
+                return this.grpcAuthHelper.checkAuthorizationAndCreateToken(streamResource,
+                        AuthHandler.Permissions.READ);
+            }
+
+            // The resource string that'll be used in the delegation token for use of the segment store
+            final String tokenResource;
+
+            // The operation that'll be specified as granted for the resource in the token. The bearer of the token
+            // will be allowed to perform the specified operation.
+            final AuthHandler.Permissions tokenPermission;
+
+            if (authParams.isStreamUserDefined()) {
+                // The operation itself requires the caller to possess read permissions.
+                AuthHandler.Permissions minimumPermissions = AuthHandler.Permissions.READ;
+
+                // This is the permission that the client is requesting to be assigned on the delegation token.
+                AuthHandler.Permissions requestedPermissions = PermissionsHelper.parse(request.getRequestedPermission(),
+                        AuthHandler.Permissions.READ);
```

Comment:
I've significantly modified the code now. Please take a look at the revised code. Older clients will now be authorized for `READ_UPDATE` as earlier so that both read and write paths will continue to work despite them not sending requested permission.